src/s3ql/http.py

```diff
@@ -0,0 +1,1558 @@
+
+        log.debug('done (in_remaining=%s)', self._in_remaining)
+        return HTTPResponse(method, path, status, reason, header, body_length)
+
+    def _setup_read(self, method, status, header):
+        '''Prepare for reading response body
+
+        Sets up `._encoding`, `_in_remaining` and returns Content-Length
+        (if available).
+
+        See RFC 2616, sec. 4.4 for specific rules.
+        '''
+
+        # On error, the exception is stored in _encoding and raised on
+        # the next call to co_read() et al - that way we can still
+        # return the http status and headers.
+
+        will_close = header.get('Connection', 'keep-alive').lower() == 'close'
+
+        body_length = header['Content-Length']
+        if body_length is not None:
+            try:
+                body_length = int(body_length)
+            except ValueError:
+                self._encoding = InvalidResponse('Invalid content-length: %s' % body_length)
+                self._in_remaining = RESPONSE_BODY_ERROR
+                return None
+
+        if (
+            status == NO_CONTENT
+            or status == NOT_MODIFIED
+            or 100 <= status < 200
+            or method == 'HEAD'
+        ):
+            log.debug('no content by RFC')
+            self._in_remaining = None
+            self._encoding = None
+            return 0
+
+        tc = header.get('Transfer-Encoding', 'identity').lower()
+        if tc == 'chunked':
+            log.debug('Chunked encoding detected')
+            self._encoding = Encodings.CHUNKED
+            self._in_remaining = 0
+            return None
+
+        elif tc != 'identity':
+            log.warning('Server uses invalid response encoding "%s"', tc)
+            self._encoding = InvalidResponse('Cannot handle %s encoding' % tc)
+            self._in_remaining = RESPONSE_BODY_ERROR
+            return None
+
+        log.debug('identity encoding detected')
+        self._encoding = Encodings.IDENTITY
+
+        if body_length is not None:
+            log.debug('Will read response body of %d bytes', body_length)
+            self._in_remaining = body_length or None
+            return body_length
+
+        if will_close:
+            log.debug('no content-length, will read until EOF')
+            self._in_remaining = READ_UNTIL_EOF
+            return None
+
+        log.debug('no content length and no chunked encoding, will raise on read')
+        self._encoding = UnsupportedResponse('No content-length and no chunked encoding')
+        self._in_remaining = RESPONSE_BODY_ERROR
+        return None
+
+    async def _co_read_status(self):
+        '''Read response line'''
+
+        log.debug('start')
+
+        # read status
+        try:
+            line = await self._co_readstr_until(b'\r\n', MAX_LINE_SIZE)
+        except _ChunkTooLong:
+            raise InvalidResponse('server send ridicously long status line')
+
+        try:
+            version, status, reason = line.split(None, 2)
```

Comment:
Would this be easier to read with a kwarg?

```suggestion
            version, status, reason = line.split(maxsplit=2)
```